src/stan/callbacks/json_writer.hpp

```diff
@@ -0,0 +1,375 @@
+   * @param key Name of the value pair
+   * @param value `std::size_t` to write.
+   */
+  void write(const std::string& key, std::size_t value) {
+    write_sep();
+    write_key(key);
+    *output_ << value;
+  }
+
+  /**
+   * Write a key-value pair where the value is a double.
+   * @param key Name of the value pair
+   * @param value double to write.
+   */
+  void write(const std::string& key, double value) {
+    write_sep();
+    write_key(key);
+    *output_ << value;
+  }
+
+  /**
+   * Write a key-value pair where the value is a complex value.
+   * @param key Name of the value pair
+   * @param value complex value to write.
+   */
+  void write(const std::string& key, const std::complex<double>& value) {
+    write_sep();
+    write_key(key);
+    *output_ << "\"" << key << "\" : [" << value.real() << ", " << value.imag()
+             << "]";
+  }
+
+  /**
+   * Write a key-value pair where the value is a vector to be made a list.
+   * @param key Name of the value pair
+   * @param values vector to write.
+   */
+  void write(const std::string& key, const std::vector<double>& values) {
+    write_sep();
+    write_key(key);
+    write_vector(values);
+  }
+
+  /**
+   * Write a key-value pair where the value is a vector of strings to be made a
+   * list.
+   * @param key Name of the value pair
+   * @param values vector of strings to write.
+   */
+  void write(const std::string& key, const std::vector<std::string>& values) {
+    write_sep();
+    write_key(key);
+    write_vector(values);
+  }
+
+  /**
+   * Write a key-value pair where the value is an Eigen Matrix.
+   * @param key Name of the value pair
+   * @param mat Eigen Matrix to write.
+   */
+  void write(const std::string& key, const Eigen::MatrixXd& mat) {
+    write_sep();
+    write_key(key);
+    Eigen::IOFormat json_format(Eigen::StreamPrecision, Eigen::DontAlignCols,
+                                ", ", ", ", "[", "]", "[", "]");
+    *output_ << mat.format(json_format);
+  }
+
+  /**
+   * Write a key-value pair where the value is an Eigen Vector.
+   * @param key Name of the value pair
+   * @param vec Eigen Vector to write.
+   */
+  void write(const std::string& key, const Eigen::VectorXd& vec) {
+    write_sep();
+    write_key(key);
+    Eigen::IOFormat json_format(Eigen::StreamPrecision, Eigen::DontAlignCols,
+                                ", ", "", "", "", "[", "]");
+    *output_ << vec.transpose().format(json_format);
+  }
+
+  /**
+   * Write a key-value pair where the value is a Eigen RowVector.
+   * @param key Name of the value pair
+   * @param vec Eigen RowVector to write.
+   */
+  void write(const std::string& key, const Eigen::RowVectorXd& vec) {
+    write_sep();
+    write_key(key);
+    Eigen::IOFormat json_format(Eigen::StreamPrecision, Eigen::DontAlignCols,
+                                ", ", "", "", "", "[", "]");
+    *output_ << vec.format(json_format);
+  }
+
+  /**
+   * Reset state
+   */
+  void reset() {
```

Comment:
What is the use case for this function?

Reply:
needed if the caller is reusing the json writer and also writing to the output stream.
right now it's used by the unit tests

Reply:
It's not needed in the tests, each test is run on a newly instantiated object (see the [FAQ](https://fuchsia.googlesource.com/third_party/googletest/+/HEAD/googletest/docs/faq.md#ctorvssetup)).

I also don't think we should support re-using the same writer on a new file, it just seems like a recipe for problems, and constructing these isn't particularly expensive. I think `reset` should not be part of the API of this